### PR TITLE
🌊 - Kim

### PR DESCRIPTION
**Pull request by kimvitug05**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | The Model manages the data of the application. It performs the business logic.
Describe in your own words what the Controller is doing in Rails | The Controller parses requests from the browser, fetches data from the Model and View and returns the data and response to the user.
Describe in your own words what the View is doing in Rails | The View is what the user sees. 
Describe an edge-case controller test you wrote | If a user tries to mark complete a nonexistent task, respond with `:not_found`
What is the purpose of using strong params? (i.e. the params method in the controller) | Strong params allow us to DRY up our code. It also adds a level of security of data sent through forms.
How are Rails migrations related to Rails models? | A migration is a change to the database schema. The command `rails generate model book title:string author:string` generates both a model and a migration. The migration provides instructions to the database on how to store the data the model is describing. 
Describe one area of Rails that are still unclear on | Partials.

### CODE REVIEW

File: app/views/tasks/edit.html.erb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+        <%= f.label :name %>
+        <%= f.text_field :name, class: "form-control" %>
+      </div>
+      <div class="form-group w-75 p-3">
+        <%= f.label :description %>
+        <%= f.text_area :description, class: "form-control" %>
+      </div>
+      <div class="form-group w-75 p-3">
+        <%= f.label :completed_at %>
+        <%= f.text_field :completed_at, class: "form-control" %>
```
Comment (dHelmgren) [inline app/views/tasks/edit.html.erb:13]: because you give this form field to the user, when someone first creates a task it has a value of `''` rather than `nil`. Therefore, when a user first goes to mark a task as complete, they end up changing the task's `completed_at` value from `''` to `nil`.

Either remove this field so that the only way that `completed_at` changes is thru the `mark_complete` route, or update your view to avoid using `.empty?` to check and instead use `.nil?`

File: app/views/tasks/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    </div>
+    <div id="tasks" class="col">
+      <h3>Your Tasks</h3>
+      <% if @tasks.length > 0 %>
+        <ol>
+          <% @tasks.each do |task| %>
+            <li>
+              <%= link_to task.name, task_path(task), :class => (task.completed_at.nil? || task.completed_at.empty? ? "" : "completed-task") %>
+              <p>
+                <%= link_to (task.completed_at.nil? || task.completed_at.empty? ? "Mark Complete" : "Mark Incomplete"), toggle_complete_task_path(task), :method => :patch %> |
```
Comment (dHelmgren) [inline app/views/tasks/index.html.erb:15]: `.empty?` returns true for `nil` and `""`.

You probably want `.nil?`.

**Review by dHelmgren — COMMENTED** [review]
# Task List




## Major Learning Goals/Code Review





| Criteria                                           | yes/no, and optionally any details/lines of code to reference |
| -------------------------------------------------- | ------------------------------------------------------------- |
| At least 6 commits with meaningful commit messages | ✔️                                                            |
| Routes follow RESTful conventions                  | ✔️                                                            |
| Uses named routes (like `_path`)                   | ✔️                                                           |
| Creates Models and migrations                      | ✔️                                                            |
| Creates styled views                               | ✔️                                                           |
| Handles errors like nonexistant tasks              | ✔️                                                            |
| Uses `form_with` to render forms in Rails          | ✔️                                                          |

## Functional Requirements/Manual Testing

| Functional Requirement                                                         | yes/no |
| ------------------------------------------------------------------------------ | ------ |
| Successfully handles index & show                                              | ✔️     |
| index & show tests pass                                                        | ✔️     |
| Successfully handles: New, Create                                              | ✔️     |
| New, Create tests pass                                                         | ✔️     |
| Successfully handles: Edit, Update                                             | ✔️     |
| Edit, Update tests pass with valid & invalid task ids                          | ✔️     |
| Successfully handles: Destroy, Task Complete                                   | Task complete doesn't work on the first click! see comments     |
| Tests for Destroy & Task Complete include tests for valid and invalid task ids | ✔️     |

## Overall Feedback

| Overall Feedback                | Criteria                                                                                                                                                                                      | yes/no |
| ------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------ |
| Green (Meets/Exceeds Standards) | 5+ in Code Review && 6+ in Functional Requirements                                                                                                                                            |✔️
| Yellow (Approaches Standards)   | 3+ in Code Review && 5+ in Functional Requirements, or the instructor judges that this project needs special attention                                                                        |
| Red (Not at Standard)           | 0-2 in Code Review or 0-4 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention |
